design/one-pager-managed-resource-api-design.md

```diff
@@ -246,6 +260,20 @@ type EmbeddedStructStatus struct {
 }
 ```
 
+#### Sensitive Input Fields
+
+Some cloud services can take sensitive input such as passwords, certificates or
+tokens. However, exposing those fields on the CR is not the best way to handle it.
+For such fields, the field name should be `<field-name>SecretRef` and the type
+of that field should be `SecretKeySelector`([from crossplane-runtime]). The controller
```

Comment:
One thing that immediately come to mind here is do we setup a watch on this secret such that if a user changes it then it triggers an update in the managed resource controller?

Reply:
That's an implementation detail but I don't think we need a watch for those secrets. In general, you don't really update those. For example, you can't really update the root certificate that you supplied. Not a strong opinion, but I feel like 1 minute long-wait interval we have would be sufficient.

Reply:
Gotcha, so setting these would just be part of the reference resolution chain?

Reply:
> would just be part of the reference resolution chain

No, not really. Reference resolution populates the field on the CR but we don't want that for sensitive fields. So, in `Create` and/or `Update`, the controller has to fetch the secret data and include it in the final SDK object.

Reply:
Hmm I wonder if we could somehow be more generic about how sensitive data is fetched for CR's such that a user could potentially specify a variety of ways to fetch secrets other than just from a Kubernetes `Secret`.

Reply:
In the implementation, we don't impose any mechanism like we do for reference resolving. So, they are actually free to do so in controller code. But in general, I believe as long as you're on Kubernetes level, it's safe to assume that the sensitive data is in a `Secret`.

I just looked into how Vault is used in Kubernetes but apparently, it just injects the secret values into the pods without using Kubernetes Secrets. For our case, this is simply not possible since the controller pod works for multiple user entities. So, I feel like our best bet for now is to rely on Kubernetes secrets. @negz what do you think?

Reply:
I think this seems like a solid direction for now, and we can iterate on it as community requests additional functionality :)

Reply:
I agree that secrets are the best place to start for now. If we wanted to future proof this API for other credential stores we could use the "union-with-discriminator" pattern described in https://github.com/kubernetes/enhancements/pull/926, e.g.

```yaml
fooCredentials:
  credentialStore: Secret
  secretRef:
    name: coolsecret
```

This would allow us to add `credentialStore: Vault`, etc in future. This all said, I probably would not worry about this for now given that none of our other credentials (providers, connection secrets) support anything other than secrets.